src/asset_util/src/lib.rs

```diff
@@ -49,8 +57,39 @@ impl CertifiedAssets {
         )
     }
 
-    fn witness_v1(&self, path: &str) -> HashTree {
-        let witness = self.certification_v1.witness(path.as_bytes());
+    /// Returns the [CertifiedAsset] for the given URL path and certificate version, if it exists.
+    /// If the canister also uses the `certified_data` to issue canister signatures, the caller
+    /// should provide the (pruned) signature (`sigs`) subtree.
+    pub fn certified_asset(
+        &self,
+        url_path: &str,
+        certificate_version: Option<u16>,
+        sigs_tree: Option<HashTree>,
+    ) -> Option<CertifiedAsset> {
+        assert!(url_path.starts_with('/'));
+        let certified_asset = self
+            .assets
+            .get(url_path)
+            .map(|(headers, content)| CertifiedAsset {
+                headers: headers.clone(),
+                content: content.clone(),
+            });
+        certified_asset.map(|mut certified_asset| {
+            match certificate_version {
+                Some(2) => certified_asset
+                    .headers
+                    .extend(self.certificate_headers_v2(url_path, sigs_tree)),
+                // Certification v1 is also the fallback for unknown certificate versions
```

Comment:
The problem is that originally clients did not specify a version at all. So in order to not break old clients, we have to assume that they only understand v1.
But I'll make the logic more robust, so that anything `>=2` uses version 2 and nothing or 1 use version 1. 👍

Reply:
as discussed, let's rename `certificate_version` to `max_certificate_version`, and specify in a comment that it means something like "max. certificate version expected by the caller", and that the actually returned version can be lower.